File: TcpSocketSystem_Java/src/com/tecnm/sockets/infrastructure/repository/RepositorioSesionesMemoria.java
package com.tecnm.sockets.infrastructure.repository;

import com.tecnm.sockets.domain.modelos.EstadoSesion;
import com.tecnm.sockets.domain.puertos.IRepositorioSesiones;
import java.util.Collection;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

/**
 * Implementación en memoria y thread-safe del repositorio de sesiones en Java.
 */
public final class RepositorioSesionesMemoria implements IRepositorioSesiones {

    private final ConcurrentHashMap<String, EstadoSesion> sesiones = new ConcurrentHashMap<>();

    @Override
    public void registrarSesion(EstadoSesion sesion) {
        sesiones.put(sesion.getSesionId(), sesion);
    }

    @Override
    public void actualizarSesion(EstadoSesion sesion) {
        sesiones.put(sesion.getSesionId(), sesion);
    }

    @Override
    public void finalizarSesion(String sesionId) {
        sesiones.remove(sesionId);
    }

    @Override
    public EstadoSesion obtenerPorId(String sesionId) {
        return sesiones.get(sesionId);
    }

    @Override
    public Collection<EstadoSesion> obtenerSesionesActivas() {
        return sesiones.values().stream()
                .filter(EstadoSesion::estaActiva)
                .collect(Collectors.toList());
    }

    @Override
    public int obtenerTotalSesionesActivas() {
        return sesiones.size();
    }
}

File: TcpSocketSystem_DotNet/src/Infrastructure/Repositories/RepositorioSesionesMemoria.cs
using System.Collections.Concurrent;
using TcpSocketSystem.Core.Domain.Modelos;
using TcpSocketSystem.Core.Domain.Puertos;

namespace TcpSocketSystem.Core.Infrastructure.Repositories;

/// <summary>
/// Repositorio de sesiones concurrentes en memoria con thread-safety (ConcurrentDictionary).
/// </summary>
public sealed class RepositorioSesionesMemoria : IRepositorioSesiones
{
    private readonly ConcurrentDictionary<string, EstadoSesion> _sesiones = new();

    public void RegistrarSesion(EstadoSesion sesion)
    {
        _sesiones[sesion.SesionId] = sesion;
    }

    public void ActualizarSesion(EstadoSesion sesion)
    {
        _sesiones[sesion.SesionId] = sesion;
    }

    public void FinalizarSesion(string sesionId)
    {
        _sesiones.TryRemove(sesionId, out _);
    }

    public EstadoSesion? ObtenerPorId(string sesionId)
    {
        _sesiones.TryGetValue(sesionId, out var sesion);
        return sesion;
    }

    public IReadOnlyCollection<EstadoSesion> ObtenerSesionesActivas()
    {
        return _sesiones.Values.Where(s => s.EstaActiva).ToList();
    }

    public int ObtenerTotalSesionesActivas()
    {
        return _sesiones.Count;
    }
}
